packages/calling-stateful-client/src/CallContext.ts

```diff
@@ -382,6 +386,56 @@ export class CallContext {
     });
   }
 
+  /* @conditional-compile-remove(reaction) */
+  public setReceivedReactionFromParticipant(callId: string, participantKey: string, reactionMessage: ReactionMessage | null): void {
+    let baseTimeStamp = new Date();
+    baseTimeStamp.setMonth(0);
+    baseTimeStamp.setDate(1);
+    baseTimeStamp.setHours(0, 0, 0, 0);
+  
+    let baseUnixTimestamp = Math.floor(baseTimeStamp.getTime() / 1000);
+    let currentUnixTimestamp = Math.floor(new Date().getTime() / 1000) - baseUnixTimestamp;
+    this.modifyState((draft: CallClientState) => {
+      const call = draft.calls[this._callIdHistory.latestCallId(callId)];
+      if(call) {
+        const participant = call.remoteParticipants[participantKey];
+        // We might see concurrency issue here.
+        if(participant) {
+
+          if(reactionMessage == null) {
+            participant.reaction = { shouldRender: false, reactionType: '', receivedTimeStamp: 0};
+          } else {
+            if(participant.reaction?.shouldRender && (currentUnixTimestamp - participant.reaction?.receivedTimeStamp) < 3000) {
+              participant.reaction = { shouldRender: false, reactionType: '', receivedTimeStamp: 0};
+              clearTimeout(this._timeOutId[participantKey]);
+            }
+  
+            participant.reaction = reactionMessage ? convertFromSDKToReactionState(reactionMessage.reactionType) : { shouldRender: false, reactionType: '', receivedTimeStamp: 0};
+            
+            this._timeOutId[participantKey] = setTimeout(() => {
+              clearParticipantReactionState(this, callId, participantKey);
+            }, 2170);
+          }
+        } else if (participantKey === toFlatCommunicationIdentifier(this._state.userId)) {
+
+          if(reactionMessage == null) {
+            call.reaction.localParticipantReactionPayload = { shouldRender: false, reactionType: '', receivedTimeStamp: 0};
+          } else {
+            if(call.reaction.localParticipantReactionPayload?.shouldRender && (currentUnixTimestamp - call.reaction?.localParticipantReactionPayload.receivedTimeStamp) < 3000) {
+                clearParticipantReactionState(this, callId, participantKey);
+                clearTimeout(this._timeOutId[participantKey]);
+              }
+          call.reaction.localParticipantReactionPayload = reactionMessage ? convertFromSDKToReactionState(reactionMessage.reactionType) : { shouldRender: false, reactionType: '', receivedTimeStamp: 0};
+          this._timeOutId[participantKey] = setTimeout( () => {
```

Comment:
timeout logic will need moved to the composite/component level

Reply:
We'll keep it here for now.